molten-core/src/main/java/com/hotels/molten/core/collapser/FanOutRequestCollapser.java

```diff
@@ -113,14 +114,15 @@ private FanOutRequestCollapser(Builder<CONTEXT, VALUE> builder) {
                 List<CONTEXT> contexts = toContexts(contextsWithSubjects);
                 LOG.debug("Executing batch with contexts={}, groupId={}", contexts, groupId);
                 updateMetrics(contextsWithSubjects);
+                final List<ContextWithSubject<CONTEXT, VALUE>> contextsWithSubjectsToDrain = new LinkedList<>(contextsWithSubjects);
                 return Mono.just(contexts)
-                    .flatMap(builder.bulkProvider)
+                    .flatMapMany(builder.bulkProvider)
                     .subscribeOn(builder.batchScheduler)
+                    .flatMap(value -> bindValueToContext(contextsWithSubjectsToDrain, value, groupId))
+                    .concatWith(bindEmptyToRemaining(contextsWithSubjectsToDrain))
                     .transform(BulkheadOperator.of(bulkhead))
```

Comment:
Maybe I'm missing something but wouldn't moving the bulkhead operator after the flatMapping of each value makes it limit the concurrency of the individual values rather than the batches?

Reply:
Yes, that's what came to my mind first as well, but no, bulkhead counts and limits subscriptions. So as we are subscribing to a `Flux` only once just like to a `Mono`, it limits the bulk calls not the individuals.

Reply:
It would be a good idea to alter the related test cases to fail if individual values would be limited, will do once. I'm at a computer again 🙂

Reply:
Added test case in 74534894.